library/Freckle/App/Datadog.hs

```diff
@@ -76,9 +85,85 @@ counter
   -> [(Text, Text)]
   -> Int
   -> m ()
-counter name tags = sendAppMetricWithTags name tags Counter
+counter name tags = sendAppMetricWithTags name tags Datadog.Counter
+
+-- | A data type containing all reporting values for a gauge
+data Gauge = Gauge
+  { gaugeName :: Text
+  , gaugeTags :: [(Text, Text)]
+  , gaugeAtomic :: EKG.Gauge
+  }
+
+-- | Create a gauge holding in memory state
+newGauge :: (MonadIO m) => Text -> [(Text, Text)] -> m Gauge
+newGauge name tags = Gauge name tags <$> liftIO EKG.new
+
+-- | Increment gauge state and report its current value
+incrementGauge
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , HasDogStatsClient env
+     , HasDogStatsTags env
+     )
+  => Gauge
+  -> m ()
+incrementGauge = addGauge 1
+
+-- | Add to gauge state and report its current value
+addGauge
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , HasDogStatsClient env
+     , HasDogStatsTags env
+     )
+  => Int64
+  -> Gauge
+  -> m ()
+addGauge i = withGauge (`EKG.add` i)
+
+-- | Decrement gauge state and report its current value
+decrementGauge
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , HasDogStatsClient env
+     , HasDogStatsTags env
+     )
+  => Gauge
+  -> m ()
+decrementGauge = subtractGauge 1
+
+-- | Subtract from gauge state and report its current value
+subtractGauge
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , HasDogStatsClient env
+     , HasDogStatsTags env
+     )
+  => Int64
+  -> Gauge
+  -> m ()
+subtractGauge i = withGauge (`EKG.subtract` i)
+
+withGauge
+  :: ( MonadUnliftIO m
+     , MonadReader env m
+     , HasDogStatsClient env
+     , HasDogStatsTags env
+     )
+  => (EKG.Gauge -> IO ())
+  -> Gauge
+  -> m ()
+withGauge f Gauge {..} = do
+  current <- liftIO $ do
+    f gaugeAtomic
+    EKG.read gaugeAtomic
+  gauge gaugeName gaugeTags $ fromIntegral current
```

Comment:
Should this be structured to be more exception safe (i.e. put `gauge` in a `finally`)?

I'm actually not sure what behavior we'd want, or if we even need to care.

Reply:
We don't have to worry about exceptions here unless something went absolutely wrong. `Atomic` is just a mutable `CInt`.